compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/CollectionStubMethodLowering.kt

```diff
@@ -0,0 +1,211 @@
+import org.jetbrains.kotlin.backend.jvm.JvmBackendContext
+import org.jetbrains.kotlin.backend.jvm.util.eraseTypeParameters
+import org.jetbrains.kotlin.descriptors.Modality
+import org.jetbrains.kotlin.ir.UNDEFINED_OFFSET
+import org.jetbrains.kotlin.ir.builders.declarations.buildFun
+import org.jetbrains.kotlin.ir.builders.irBlockBody
+import org.jetbrains.kotlin.ir.builders.irCall
+import org.jetbrains.kotlin.ir.builders.irString
+import org.jetbrains.kotlin.ir.declarations.*
+import org.jetbrains.kotlin.ir.declarations.impl.IrValueParameterImpl
+import org.jetbrains.kotlin.ir.symbols.IrClassSymbol
+import org.jetbrains.kotlin.ir.symbols.IrTypeParameterSymbol
+import org.jetbrains.kotlin.ir.symbols.impl.IrValueParameterSymbolImpl
+import org.jetbrains.kotlin.ir.types.*
+import org.jetbrains.kotlin.ir.util.*
+import org.jetbrains.kotlin.name.FqName
+
+internal val collectionStubMethodLowering = makeIrFilePhase(
+    ::CollectionStubMethodLowering,
+    name = "CollectionStubMethod",
+    description = "Generate Collection stub methods"
+)
+
+private class CollectionStubMethodLowering(val context: JvmBackendContext) : ClassLoweringPass {
+
+    private val state = context.state
+
+    private val typeMapper = state.typeMapper
+
+    override fun lower(irClass: IrClass) {
+        if (irClass.isInterface) {
+            return
+        }
+
+        val methodStubsToGenerate = findRelevantStubMethods(irClass)
+
+        // We don't need to generate stub for existing methods, but for FAKE_OVERRIDE methods with ABSTRACT modality,
+        // it means an abstract function in superclass that is not implemented yet,
+        // stub generation is still needed to avoid invocation error.
+        val existingMethodSignatures = irClass.functions.filterNot {
+            it.modality == Modality.ABSTRACT && it.origin == IrDeclarationOrigin.FAKE_OVERRIDE
+        }.associateBy { it.toSignature() }
+
+        for (member in methodStubsToGenerate) {
+            if (member.toSignature() in existingMethodSignatures.keys) {
+                // In the case that we find a defined method that matches the stub signature, we add the overridden symbols to that
+                // defined method, so that bridge lowering can still generate correct bridge for that method
+                existingMethodSignatures[member.toSignature()]?.overriddenSymbols?.addAll(
+                    member.overriddenSymbols
+                )
+            } else {
+                irClass.declarations.add(member)
+            }
+        }
+    }
+
+    //TODO: replace with new typeMapper using no descriptor
+    private fun IrSimpleFunction.toSignature() = typeMapper.mapAsmMethod(this.descriptor).toString()
+
+    private fun createStubMethod(function: IrSimpleFunction, irClass: IrClass, typeMap: Map<IrTypeParameterSymbol, IrType>): IrSimpleFunction {
+        return buildFun {
+            name = function.name
+            returnType = function.returnType.substitute(typeMap)
```

Comment:
Shouldn't we call `eraseTypeParameters` here as well? Or is it not relevant because none of the methods in mutable collection classes have return types with generic type arguments?

Reply:
It seems to be not relevant, but I extracted the type erasure logic from bridgelowering when I was using erasure in stub lowering, do you want me to put it back? Or we can keep it as it might be used in other occasions?

Reply:
Well I think I should put it back, since Steven is having a [PR](https://github.com/JetBrains/kotlin/pull/2339) for extracting erasure from bridge lowering already.